### PR TITLE
[addons][inputstream][videocodec] API improvement

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description

This changes are taken from https://github.com/xbmc/xbmc/pull/17438 and to bring the more less critical changes on another request and have to other smaller and more clean.

## User related:

Not user relevant

Only for all binary addon system requests may be for the user:
- Binary addon system revised to ensure higher security in its data exchange between addon and Kodi

Commit 1:
--------------------
**[addons][inputstream] make all "C" interface structures to own memory**

This done to prevent API backward problems if something becomes replaced e.g. on props or to kodi calls.

Commit 2:
--------------------
**[addons][inputstream] use "C" defines about array sizes**

There was before this values set direct by number and by MAX_STREAM_COUNT as static const value within structure.

Here becomes changed to use them as "C" defines, this needed to have a safe interface between Kodi and addon. Also to allow "C" only for other languages.

Commit 3:
--------------------
**[addons][codec] rename buffer to videoBufferHandle on VIDEOCODEC_PICTURE**

Before with named "buffer" was for me on begin confusing and thought first
about data buffer used on addons itself. This was only a pointer to a
class in Kodi itself and not handled by addons.

Commit 4:
--------------------
**[addons][codec] use every addon functions structure in own mem**

Before was them all together in one memory where makes it with API harder if something was added.

Commit 5:
--------------------
**[addons][codec] don't use C++ class pointer within "C" structure**

This makes it hard to use the header for "C" only (e.g. base for other languages).

Commit 6:
--------------------
**[addons][inputstream] don't use C++ shared_ptr within "C" interface**

There was a shared_ptr used on DemuxPacket where can be critical if addon and Kodi was compiled with different versions.

On following changes are the DemuxPacket and DemuxCryptoInfo also cleaned to match "C".

Commit 7:
--------------------
**[addons][binary] increase inputstream and videocodec API version**

About inputstream becomes increased to 2.3.4
About videocodec becomes increased to 1.0.4

Increase related to last changes where addons need a recompile.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/Inputstream.h [hunk 1]
```diff
@@ -238,17 +242,20 @@ extern "C"
 
     uint32_t m_flags;
 
-    char m_name[256]; /*!< @brief (optinal) name of the stream, \0 for default handling */
-    char m_codecName[32]; /*!< @brief (required) name of codec according to ffmpeg */
+    char m_name
+        [INPUTSTREAM_MAX_STRING_NAME_SIZE]; /*!< @brief (optinal) name of the stream, \0 for default handling */
+    char m_codecName
+        [INPUTSTREAM_MAX_STRING_CODEC_SIZE]; /*!< @brief (required) name of codec according to ffmpeg */
```
Comment (garbear) [inline xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/Inputstream.h:248]: You could avoid this awkward spacing by moving the doxy to before the symbol

```c++
//! @brief (optional) name of the stream, \0 for default handling
char m_name[INPUTSTREAM_MAX_STRING_NAME_SIZE];

//! @brief (required) name of codec according to ffmpeg
char m_codecName[INPUTSTREAM_MAX_STRING_CODEC_SIZE];
```

Or however the doxy is best formatted. Also: optinal -> optional

**Review by garbear — APPROVED** [review]
I reviewed the 7 commits and all are good improvements. No non-cosmetic issues jump out at me.

File: xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/Inputstream.h (hunk 1 above)
Comment (AlwinEsch) [reply xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/Inputstream.h]: Have changed it, yes much better then as clang not do his line breaks inside.

**Comment by lrusak** [comment]
Not sure if you noticed but it seems that jenkins keeps failing because it see's a new warning for some reason. As far as I can tell though the change isn't introduced in this PR but this PR seems to trip up jenkins. 

https://jenkins.kodi.tv/job/Android-ARM/lastFailedBuild/clang/new/source.9c81b08d-08a8-466b-8649-416874a06ea9/#62
https://jenkins.kodi.tv/job/Android-ARM/lastFailedBuild/clang/new/source.39aba585-7188-462b-925f-1dc4833ade34/#208

```
09:01:57 /home/jenkins/workspace/Android-ARM/xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp:62:15: warning: braces around scalar initializer [-Wbraced-scalar-init]
09:01:57   m_struct = {{ 0 }};
09:01:57               ^~~~~
09:01:57 /home/jenkins/workspace/Android-ARM/xbmc/cores/VideoPlayer/DVDInputStreams/InputStreamAddon.cpp:208:15: warning: braces around scalar initializer [-Wbraced-scalar-init]
09:01:57   m_struct = {{ 0 }};
09:01:57               ^~~~~
09:01:57 2 warnings generated.
```

**Comment by AlwinEsch** [comment]
Not seen, thanks a lot for report about!

Change and fix it soon.

**Comment by AlwinEsch** [comment]
@peak3d what's the current state about GSOC?

About inputstream/videocodec, there are the last places with cleanup about safe "C" ABI interface :smile:

**Comment by AlwinEsch** [comment]
@peak3d @phunkyfish I bring this in on monday 12. October, as we can not wait any longer to get the inputstream parts done.

**Comment by AlwinEsch** [comment]
Is now in, tomorrow after the Ubuntu launchpad about the Kodi is finished can be requests on addons in.

**Pull request merged**